File: frontend/app/utils/captionText.js
const DEVANAGARI = /[\u0900-\u097F]/g;
const LATIN = /[A-Za-z]/g;
const JUNK = /^(?:[|¦]+|♪+|♫+|…+|\.{2,}|[-–—]+|\/+|\\+|\[(?:music|applause|laughter|inaudible)\]|\((?:music|applause)\))$/i;

export function isJunkToken(token) {
  const t = String(token || '').trim();
  if (!t) return true;
  return JUNK.test(t);
}

export function sanitizeCaptionText(text) {
  if (!text || typeof text !== 'string') return '';

  let out = text.replace(/\u00a0/g, ' ').replace(/[|¦]/g, ' ');
  out = out.replace(/[♪♫]+/g, ' ');
  out = out.replace(/\[(?:music|applause|laughter|inaudible)\]/gi, ' ');
  out = out.replace(/\((?:music|applause)\)/gi, ' ');

  out = out.split(/\s+/).filter((tok) => !isJunkToken(tok)).join(' ');

  const letters = (out.match(DEVANAGARI) || []).length + (out.match(LATIN) || []).length;
  const dev = (out.match(DEVANAGARI) || []).length;
  const isDev = letters > 0 && dev / letters >= 0.3;

  if (!isDev) {
    out = out.replace(/।/g, '.');
    out = out.replace(/\.{2,}/g, '.');
    out = out.replace(/\s+([.,!?;:])/g, '$1');
    out = out.replace(/^[.,;:]+|[.,;:]+$/g, '');
  }

  return out.replace(/\s+/g, ' ').trim();
}

export function sanitizeSubtitleList(subtitles = []) {
  return (subtitles || [])
    .map((sub) => {
      const words = (sub.words || [])
        .map((w) => ({ ...w, text: sanitizeCaptionText(w.text || '') }))
        .filter((w) => w.text && !isJunkToken(w.text));
      const text = sanitizeCaptionText(sub.text || '') || words.map((w) => w.text).join(' ');
      return { ...sub, text, words };
    })
    .filter((sub) => sub.text);
}

File: backend/utils/caption_text.py
"""Caption text invariants: language-aware cleanup of Whisper artifacts."""

from __future__ import annotations

import re
import unicodedata

DEVANAGARI_RE = re.compile(r"[\u0900-\u097F]")
LATIN_RE = re.compile(r"[A-Za-z]")
JUNK_TOKEN_RE = re.compile(
    r"^(?:[|¦]+|♪+|♫+|…+|\.{2,}|[-–—]+|/+|\\+|\[(?:music|applause|laughter|inaudible)\]|\((?:music|applause)\))$",
    re.IGNORECASE,
)


def is_devanagari_text(text: str) -> bool:
    if not text:
        return False
    letters = DEVANAGARI_RE.findall(text) + LATIN_RE.findall(text)
    if not letters:
        return False
    dev = len(DEVANAGARI_RE.findall(text))
    return (dev / len(letters)) >= 0.3


def is_junk_token(token: str) -> bool:
    t = (token or "").strip()
    if not t:
        return True
    return bool(JUNK_TOKEN_RE.match(t))


def sanitize_caption_text(text: str) -> str:
    """Strip Whisper junk and keep Nepali danda only on Devanagari text."""
    if not text or not isinstance(text, str):
        return ""

    text = unicodedata.normalize("NFC", text)
    text = text.replace("\u00a0", " ")
    text = text.replace("|", " ")
    text = text.replace("¦", " ")
    text = re.sub(r"[♪♫]+", " ", text)
    text = re.sub(r"\[(?:music|applause|laughter|inaudible)\]", " ", text, flags=re.IGNORECASE)
    text = re.sub(r"\((?:music|applause)\)", " ", text, flags=re.IGNORECASE)

    tokens = [tok for tok in text.split() if not is_junk_token(tok)]
    text = " ".join(tokens)

    if not is_devanagari_text(text):
        text = text.replace("।", ".")
        text = re.sub(r"\.{2,}", ".", text)
        text = re.sub(r"\s+([.,!?;:])", r"\1", text)
        text = re.sub(r"^[.,;:]+|[.,;:]+$", "", text)

    text = re.sub(r"\s+", " ", text).strip()
    return text
